File: Pictograph/Models/CurrentUser.m
//
//  CurrentUser.m
//  Pictograph
//
//  Created by Adam Boyd on 2015-10-19.
//  Copyright © 2015 Adam Boyd. All rights reserved.
//

#import "CurrentUser.h"

#define kFirstTimeOpeningAppKey @"kFirstTimeOpeningAppKey"
#define kEncryptionEnabledKey @"kEncryptionEnabledKey"
#define kEncryptionKey @"kEncryptionKey"

@implementation CurrentUser

@synthesize firstTimeOpeningApp;
@synthesize encryptionEnabled;
@synthesize encryptionKey;

- (id)init {
    if (self = [super init]) {
        firstTimeOpeningApp = YES;
        encryptionEnabled = NO;
        encryptionKey = @"";
    }
    
    return self;
}

#pragma mark NSCoding

- (id)initWithCoder:(NSCoder *)aDecoder {
    //We saved the bool as an NSNumber because BOOL types are not savable with an NSCoder
    NSNumber *firstTimeOpeningAppNumber = [aDecoder decodeObjectOfClass:[NSNumber class] forKey:kFirstTimeOpeningAppKey];
    firstTimeOpeningApp = [firstTimeOpeningAppNumber boolValue];
    
    NSNumber *encryptionEnabledNumber = [aDecoder decodeObjectOfClass:[NSNumber class] forKey:kEncryptionEnabledKey];
    encryptionEnabled = [encryptionEnabledNumber boolValue];
    
    encryptionKey = [aDecoder decodeObjectOfClass:[NSString class] forKey:kEncryptionKey];
    return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder {
    //First, convert the bool to an NSNumber
    NSNumber *firstTimeOpeningAppNumber = [[NSNumber alloc] initWithBool:firstTimeOpeningApp];
    [aCoder encodeObject:firstTimeOpeningAppNumber forKey:kFirstTimeOpeningAppKey];
    
    NSNumber *encryptionEnabledNumber = [[NSNumber alloc] initWithBool:encryptionEnabled];
    [aCoder encodeObject:encryptionEnabledNumber forKey:kEncryptionEnabledKey];
    
    [aCoder encodeObject:encryptionKey forKey:kEncryptionKey];
}

#pragma mark NSSecureCoding
+ (BOOL)supportsSecureCoding {
    return YES;
}

@end

File: Pictograph/Models/CurrentUser.swift
//
//  CurrentUser.swift
//  Pictograph
//
//  Created by Adam Boyd on 2015-10-25.
//  Copyright © 2015 Adam Boyd. All rights reserved.
//

import Foundation

//Need to use hungarian notation with the keys because that is how it was originally, it would be possible to change it later
private let firstTimeOpeningAppKey = "kFirstTimeOpeningAppKey"
private let encryptionEnabledKey = "kEncryptionEnabledKey"
private let encryptionPasswordKey = "kEncryptionKey"
private let showPasswordOnScreenKey = "showPasswordOnScreenKey"
private let nightModeEnabledKey = "nightModeEnabledKey"

class CurrentUser: NSObject, NSCoding, NSSecureCoding {
    var firstTimeOpeningApp: Bool = true
    var encryptionEnabled: Bool = false
    var encryptionPassword : NSString = ""
    var showPasswordOnScreen: Bool = true
    var nightModeEnabled: Bool = false
    
    override init() { super.init() }
    
    //MARK: - NSCoding
    required init?(coder aDecoder: NSCoder) {
        //Variables in from start
        
        //First time opening app
        if let firstTimeOpeningAppNumber = aDecoder.decodeObjectOfClass(NSNumber.self, forKey: firstTimeOpeningAppKey) {
            firstTimeOpeningApp = firstTimeOpeningAppNumber.boolValue
        }
        
        //Encryption enabled
        if let encryptionEnabledNumber = aDecoder.decodeObjectOfClass(NSNumber.self, forKey: encryptionEnabledKey) {
            encryptionEnabled = encryptionEnabledNumber.boolValue
        }
        
        //Encryption key
        if let storedEncryptionPassword = aDecoder.decodeObjectOfClass(NSString.self, forKey: encryptionPasswordKey) {
            encryptionPassword = storedEncryptionPassword
        }
        
        
        //Variables added in 1.1
        
        //Showing the password on screen
        if let showPasswordOnScreenNumber = aDecoder.decodeObjectOfClass(NSNumber.self, forKey: showPasswordOnScreenKey) {
            showPasswordOnScreen = showPasswordOnScreenNumber.boolValue
        }
        
        //Variables added in 1.2
        
        //If night mode is enabled
        if let nightModeEnabledNumber = aDecoder.decodeObjectOfClass(NSNumber.self, forKey: nightModeEnabledKey) {
            nightModeEnabled = nightModeEnabledNumber.boolValue
        }
        
    }
    
    func encodeWithCoder(aCoder: NSCoder) {
        let firstTimeOpeningAppNumber = NSNumber(bool: firstTimeOpeningApp)
        aCoder.encodeObject(firstTimeOpeningAppNumber, forKey: firstTimeOpeningAppKey)
        
        let encryptionEnabledNumber = NSNumber(bool: encryptionEnabled)
        aCoder.encodeObject(encryptionEnabledNumber, forKey: encryptionEnabledKey)
        
        aCoder.encodeObject(encryptionPassword, forKey: encryptionPasswordKey)
        
        let showPasswordOnScreenNumber = NSNumber(bool: showPasswordOnScreen)
        aCoder.encodeObject(showPasswordOnScreenNumber, forKey: showPasswordOnScreenKey)
        
        let nightModeEnabledNumber = NSNumber(bool: nightModeEnabled)
        aCoder.encodeObject(nightModeEnabledNumber, forKey: nightModeEnabledKey)
    }
    
    //MARK: - NSSecureCoding
    class func supportsSecureCoding() -> Bool {
        return true
    }
}